lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs

```diff
@@ -0,0 +1,575 @@
+-- how to perform a selection.
+--
+module Cardano.Wallet.Primitive.CoinSelection.MA.RoundRobin
+    (
+    -- * Performing a selection
+      performSelection
+    , SelectionCriteria (..)
+    , SelectionResult (..)
+    , SelectionError (..)
+    , BalanceInsufficientError (..)
+
+    -- * Running a selection (without making change)
+    , runSelection
+    , SelectionState (..)
+
+    -- * Running a selection step
+    , runSelectionStep
+    , SelectionLens (..)
+
+    -- * Making change
+    , makeChange
+    , makeChangeForCoin
+    , makeChangeForPaymentAssets
+    , makeChangeForSurplusAssets
+
+    -- * Partitioning
+    , partitionCoin
+    , partitionTokenQuantity
+    , partitionValue
+
+    -- * Grouping and ungrouping
+    , groupByKey
+    , ungroupByKey
+
+    -- * Round-robin processing
+    , runRoundRobin
+    , runRoundRobinM
+
+    -- * Utility functions
+    , distance
+
+    ) where
+
+import Prelude
+
+import Algebra.PartialOrd
+    ( PartialOrd (..) )
+import Cardano.Numeric.Util
+    ( padCoalesce, partitionNatural )
+import Cardano.Wallet.Primitive.Types.Coin
+    ( Coin (..) )
+import Cardano.Wallet.Primitive.Types.TokenBundle
+    ( TokenBundle )
+import Cardano.Wallet.Primitive.Types.TokenMap
+    ( AssetId, TokenMap )
+import Cardano.Wallet.Primitive.Types.TokenQuantity
+    ( TokenQuantity (..) )
+import Cardano.Wallet.Primitive.Types.Tx
+    ( TxIn, TxOut )
+import Cardano.Wallet.Primitive.Types.UTxOIndex
+    ( SelectionFilter (..), UTxOIndex (..) )
+import Control.Monad.Random.Class
+    ( MonadRandom (..) )
+import Data.Function
+    ( (&) )
+import Data.Functor.Identity
+    ( Identity (..) )
+import Data.Generics.Internal.VL.Lens
+    ( view )
+import Data.Generics.Labels
+    ()
+import Data.List.NonEmpty
+    ( NonEmpty (..) )
+import Data.Map.Strict
+    ( Map )
+import Data.Maybe
+    ( fromMaybe )
+import Data.Set
+    ( Set )
+import GHC.Stack
+    ( HasCallStack )
+import Numeric.Natural
+    ( Natural )
+
+import qualified Cardano.Wallet.Primitive.Types.TokenBundle as TokenBundle
+import qualified Cardano.Wallet.Primitive.Types.TokenMap as TokenMap
+import qualified Cardano.Wallet.Primitive.Types.UTxOIndex as UTxOIndex
+import qualified Data.Foldable as F
+import qualified Data.List as L
+import qualified Data.List.NonEmpty as NE
+import qualified Data.Map.Strict as Map
+import qualified Data.Set as Set
+
+--------------------------------------------------------------------------------
+-- Performing a selection
+--------------------------------------------------------------------------------
+
+-- | Criteria for performing a selection.
+--
+data SelectionCriteria = SelectionCriteria
+    { outputsToCover
+        :: NonEmpty TxOut
+    , utxoAvailable
+        :: !UTxOIndex
+    }
+    deriving (Eq, Show)
+
+-- | The result of performing a successful selection.
+--
+data SelectionResult = SelectionResult
+    { inputsSelected
+        :: !(NonEmpty (TxIn, TxOut))
```

Comment:
We'll need to add (i.e. will do in my addition PR) the _modified outputs_ here too. That is, each output needs to also include a minimum Ada value. This is not typically set by users. One would say "I want to send 10 Apples to Alice", but the actual output would really be  "10 Apples & 1 Ada". In practice, it shouldn't affect users _too much_ because the extra Ada would be coming from selected inputs which contains that particular assets (and therefore, bring with them some existing ada deposit). The only user who pay the full cost of this is the one forging the initial asset supply. 

I think that the `performSelection` is the right moment to do this. As an input, one can provide the `outputsToCover` as requested by the user, but we would slightly alter each output to include a minimum coin value (unless a coin value is explicitly specified) and run the coin selection with this altered output set. I think we should also make it crystal clear in the API doc.